### PR TITLE
identity/policies: Add support for managing policies in Triton

### PR DESCRIPTION
**Pull request by stack72**
```
% [add-support-for-policies] % go test -v
=== RUN   TestListPolicies
=== RUN   TestListPolicies/successful
=== RUN   TestListPolicies/eof
=== RUN   TestListPolicies/bad_decode
=== RUN   TestListPolicies/error
--- PASS: TestListPolicies (0.00s)
    --- PASS: TestListPolicies/successful (0.00s)
    --- PASS: TestListPolicies/eof (0.00s)
    --- PASS: TestListPolicies/bad_decode (0.00s)
    --- PASS: TestListPolicies/error (0.00s)
=== RUN   TestGetPolicy
=== RUN   TestGetPolicy/successful
=== RUN   TestGetPolicy/eof
=== RUN   TestGetPolicy/bad_decode
=== RUN   TestGetPolicy/error
--- PASS: TestGetPolicy (0.00s)
    --- PASS: TestGetPolicy/successful (0.00s)
    --- PASS: TestGetPolicy/eof (0.00s)
    --- PASS: TestGetPolicy/bad_decode (0.00s)
    --- PASS: TestGetPolicy/error (0.00s)
=== RUN   TestDeletePolicy
=== RUN   TestDeletePolicy/successful
=== RUN   TestDeletePolicy/error
--- PASS: TestDeletePolicy (0.00s)
    --- PASS: TestDeletePolicy/successful (0.00s)
    --- PASS: TestDeletePolicy/error (0.00s)
=== RUN   TestUpdatePolicy
=== RUN   TestUpdatePolicy/successful
=== RUN   TestUpdatePolicy/error
--- PASS: TestUpdatePolicy (0.00s)
    --- PASS: TestUpdatePolicy/successful (0.00s)
    --- PASS: TestUpdatePolicy/error (0.00s)
=== RUN   TestCreatePolocy
=== RUN   TestCreatePolocy/successful
=== RUN   TestCreatePolocy/error
--- PASS: TestCreatePolocy (0.00s)
    --- PASS: TestCreatePolocy/successful (0.00s)
    --- PASS: TestCreatePolocy/error (0.00s)
=== RUN   TestListUsers
=== RUN   TestListUsers/successful
=== RUN   TestListUsers/eof
=== RUN   TestListUsers/bad_decode
=== RUN   TestListUsers/error
--- PASS: TestListUsers (0.00s)
    --- PASS: TestListUsers/successful (0.00s)
    --- PASS: TestListUsers/eof (0.00s)
    --- PASS: TestListUsers/bad_decode (0.00s)
    --- PASS: TestListUsers/error (0.00s)
=== RUN   TestGetUser
=== RUN   TestGetUser/successful
=== RUN   TestGetUser/eof
=== RUN   TestGetUser/bad_decode
=== RUN   TestGetUser/error
--- PASS: TestGetUser (0.00s)
    --- PASS: TestGetUser/successful (0.00s)
    --- PASS: TestGetUser/eof (0.00s)
    --- PASS: TestGetUser/bad_decode (0.00s)
    --- PASS: TestGetUser/error (0.00s)
=== RUN   TestDeleteUser
=== RUN   TestDeleteUser/successful
=== RUN   TestDeleteUser/error
--- PASS: TestDeleteUser (0.00s)
    --- PASS: TestDeleteUser/successful (0.00s)
    --- PASS: TestDeleteUser/error (0.00s)
=== RUN   TestCreateUser
=== RUN   TestCreateUser/successful
=== RUN   TestCreateUser/error
--- PASS: TestCreateUser (0.00s)
    --- PASS: TestCreateUser/successful (0.00s)
    --- PASS: TestCreateUser/error (0.00s)
=== RUN   TestUpdateUser
=== RUN   TestUpdateUser/successful
=== RUN   TestUpdateUser/error
--- PASS: TestUpdateUser (0.00s)
    --- PASS: TestUpdateUser/successful (0.00s)
    --- PASS: TestUpdateUser/error (0.00s)
PASS
ok  	github.com/joyent/triton-go/identity	0.012s
```

### CODE REVIEW

File: identity/policies.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+	ID          string   `json:"id"`
+	Name        string   `json:"name"`
+	Rules       []string `json:"rules"`
+	Description string   `json:"description"`
+}
+
+type ListPoliciesInput struct{}
+
+func (c *PoliciesClient) List(ctx context.Context, _ *ListPoliciesInput) ([]*Policy, error) {
+	path := fmt.Sprintf("/%s/policies", c.client.AccountName)
```
Comment (sean-) [inline identity/policies.go]: Not for this PR, but see #88

File: identity/policies.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
+package identity
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/hashicorp/errwrap"
```
Comment (sean-) [inline identity/policies.go]: Being removed shortly...

File: identity/policies.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+	}
+
+	return result, nil
+}
+
+type GetPolicyInput struct {
+	PolicyID string
+}
+
+func (c *PoliciesClient) Get(ctx context.Context, input *GetPolicyInput) (*Policy, error) {
```
Comment (sean-) [inline identity/policies.go]: Use PBV instead of pointer if possible

File: identity/policies.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+	}
+
+	return result, nil
+}
+
+type DeletePolicyInput struct {
+	PolicyID string
+}
+
+func (c *PoliciesClient) Delete(ctx context.Context, input *DeletePolicyInput) error {
```
Comment (sean-) [inline identity/policies.go]: PBV

File: identity/policies.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+// UpdatePolicyInput represents the options that can be specified
+// when updating a policy. Anything but ID can be modified.
+type UpdatePolicyInput struct {
+	PolicyID    string   `json:"id"`
+	Name        string   `json:"name,omitempty"`
+	Rules       []string `json:"rules,omitempty"`
+	Description string   `json:"description,omitempty"`
+}
+
+func (c *PoliciesClient) Update(ctx context.Context, input *UpdatePolicyInput) (*Policy, error) {
```
Comment (sean-) [inline identity/policies.go]: PBV

File: identity/policies.go [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+	return result, nil
+}
+
+type CreatePolicyInput struct {
+	Name        string   `json:"name"`
+	Rules       []string `json:"rules"`
+	Description string   `json:"description,omitempty"`
+}
+
+func (c *PoliciesClient) Create(ctx context.Context, input *CreatePolicyInput) (*Policy, error) {
```
Comment (sean-) [inline identity/policies.go]: PBV

File: identity/policies.go [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+func (c *PoliciesClient) Update(ctx context.Context, input *UpdatePolicyInput) (*Policy, error) {
+	path := fmt.Sprintf("/%s/policies/%s", c.client.AccountName, input.PolicyID)
+	reqInputs := client.RequestInput{
+		Method: http.MethodPost,
+		Path:   path,
+		Body:   input,
+	}
+	respReader, err := c.client.ExecuteRequest(ctx, reqInputs)
+	if respReader != nil {
```
Comment (sean-) [inline identity/policies.go]: We can move this to after the `err != nil`.  `ExecuteRequest()` is guaranteed to not need the response closed if an error is returned.  This makes the code more idiomatic.

File: identity/policies.go [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+func (c *PoliciesClient) Delete(ctx context.Context, input *DeletePolicyInput) error {
+	path := fmt.Sprintf("/%s/policies/%s", c.client.AccountName, input.PolicyID)
+	reqInputs := client.RequestInput{
+		Method: http.MethodDelete,
+		Path:   path,
+	}
+	respReader, err := c.client.ExecuteRequest(ctx, reqInputs)
+	if respReader != nil {
```
Comment (sean-) [inline identity/policies.go]: Move below `err != nil`

File: identity/policies.go [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+type ListPoliciesInput struct{}
+
+func (c *PoliciesClient) List(ctx context.Context, _ *ListPoliciesInput) ([]*Policy, error) {
+	path := fmt.Sprintf("/%s/policies", c.client.AccountName)
+	reqInputs := client.RequestInput{
+		Method: http.MethodGet,
+		Path:   path,
+	}
+	respReader, err := c.client.ExecuteRequest(ctx, reqInputs)
+	if respReader != nil {
```
Comment (sean-) [inline identity/policies.go]: Move below `err != nil`

File: identity/policies.go [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+func (c *PoliciesClient) Get(ctx context.Context, input *GetPolicyInput) (*Policy, error) {
+	path := fmt.Sprintf("/%s/policies/%s", c.client.AccountName, input.PolicyID)
+	reqInputs := client.RequestInput{
+		Method: http.MethodGet,
+		Path:   path,
+	}
+	respReader, err := c.client.ExecuteRequest(ctx, reqInputs)
+	if respReader != nil {
```
Comment (sean-) [inline identity/policies.go]: Move below `err != nil`

File: identity/policies.go [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+
+func (c *PoliciesClient) Create(ctx context.Context, input *CreatePolicyInput) (*Policy, error) {
+	path := fmt.Sprintf("/%s/policies", c.client.AccountName)
+	reqInputs := client.RequestInput{
+		Method: http.MethodPost,
+		Path:   path,
+		Body:   input,
+	}
+	respReader, err := c.client.ExecuteRequest(ctx, reqInputs)
+	if respReader != nil {
```
Comment (sean-) [inline identity/policies.go]: Move down

File: identity/policies_test.go [hunk 12]
```diff
@@ -0,0 +1,453 @@
...
+	"net/http"
+	"strings"
+	"testing"
+
+	"github.com/joyent/triton-go/identity"
+	"github.com/joyent/triton-go/testutils"
+)
+
+var (
+	listPoliciesErrorType = errors.New("Error executing ListPolicies request:")
```
Comment (sean-) [inline identity/policies_test.go]: Even though this is new and matching the style of the rest of the package, can we use this as the beachhead to start using lowercase error messages?

File: identity/policies_test.go [hunk 13]
```diff
@@ -0,0 +1,453 @@
...
+	identityClient := MockIdentityClient()
+
+	do := func(ctx context.Context, ic *identity.IdentityClient) ([]*identity.Policy, error) {
+		defer testutils.DeactivateClient()
+
+		policies, err := ic.Policies().List(ctx, &identity.ListPoliciesInput{})
+		if err != nil {
+			return nil, err
+		}
+		return policies, nil
```
Comment (sean-) [inline identity/policies_test.go]: Add newline above this `return`.

File: identity/policies_test.go [hunk 14]
```diff
@@ -0,0 +1,453 @@
...
+}
+
+func TestGetPolicy(t *testing.T) {
+	identityClient := MockIdentityClient()
+
+	do := func(ctx context.Context, ic *identity.IdentityClient) (*identity.Policy, error) {
+		defer testutils.DeactivateClient()
+
+		user, err := ic.Policies().Get(ctx, &identity.GetPolicyInput{
+			PolicyID: "95ca7b25-5c8f-4c1b-92da-4276f23807f3",
```
Comment (sean-) [inline identity/policies_test.go]: Can you add a comment explaining this magic UUID?

File: identity/policies_test.go [hunk 15]
```diff
@@ -0,0 +1,453 @@
...
+		})
+		if err != nil {
+			return nil, err
+		}
+		return user, nil
```
Comment (sean-) [inline identity/policies_test.go]: Add newline above `return`.

File: identity/policies_test.go [hunk 16]
```diff
@@ -0,0 +1,453 @@
...
+}
+
+func TestDeletePolicy(t *testing.T) {
+	identityClient := MockIdentityClient()
+
+	do := func(ctx context.Context, ic *identity.IdentityClient) error {
+		defer testutils.DeactivateClient()
+
+		return ic.Policies().Delete(ctx, &identity.DeletePolicyInput{
+			PolicyID: "8700e959-4cb3-4337-8afa-fb0a53b5366e",
```
Comment (sean-) [inline identity/policies_test.go]: Magic constant?

File: identity/policies_test.go [hunk 17]
```diff
@@ -0,0 +1,453 @@
...
+}
+
+func TestUpdatePolicy(t *testing.T) {
+	identityClient := MockIdentityClient()
+
+	do := func(ctx context.Context, ic *identity.IdentityClient) (*identity.Policy, error) {
+		defer testutils.DeactivateClient()
+
+		policy, err := ic.Policies().Update(ctx, &identity.UpdatePolicyInput{
+			PolicyID:    "95ca7b25-5c8f-4c1b-92da-4276f23807f3",
```
Comment (sean-) [inline identity/policies_test.go]: Magic constant?

File: identity/policies_test.go [hunk 18]
```diff
@@ -0,0 +1,453 @@
...
+		policy, err := ic.Policies().Create(ctx, &identity.CreatePolicyInput{
+			Name:        "Test Policy",
+			Description: "Test Description",
+			Rules:       []string{"CAN rebootmachine"},
+		})
+
+		if err != nil {
+			return nil, err
+		}
+		return policy, nil
```
Comment (sean-) [inline identity/policies_test.go]: Newline

File: identity/policies_test.go [hunk 19]
```diff
@@ -0,0 +1,453 @@
...
+  },
+  {
+    "name": "createinstance",
+    "id": "95ca7b25-5c8f-4c1b-92da-4276f23805ds",
+    "rules": [
+      "can createinstance"
+    ]
+  }
+]`)
+	return &http.Response{
```
Comment (sean-) [inline identity/policies_test.go]: Add newline

File: identity/policies_test.go [hunk 20]
```diff
@@ -0,0 +1,453 @@
...
+	body := strings.NewReader(`{[
+	{
+    "name": "createinstance",
+    "id": "95ca7b25-5c8f-4c1b-92da-4276f23805ds",
+    "rules": [
+      "can createinstance"
+    ]
+  }
+]}`)
+	return &http.Response{
```
Comment (sean-) [inline identity/policies_test.go]: Add `\n`

File: identity/policies_test.go [hunk 21]
```diff
@@ -0,0 +1,453 @@
...
+	return nil, listPoliciesErrorType
+}
+
+func getPolicySuccess(req *http.Request) (*http.Response, error) {
+	header := http.Header{}
+	header.Add("Content-Type", "application/json")
+
+	body := strings.NewReader(`{
+    "name": "readinstance",
+    "id": "95ca7b25-5c8f-4c1b-92da-4276f23807f3",
```
Comment (sean-) [inline identity/policies_test.go]: Same magic UUID?  Maybe make it a `const` + comment?

File: identity/policies_test.go [hunk 22]
```diff
@@ -0,0 +1,453 @@
...
+    "rules": [
+      "can listmachine and getmachine"
+    ]
+  }
+`)
+	return &http.Response{
```
Comment (sean-) [inline identity/policies_test.go]: Add `\n`

File: identity/policies_test.go [hunk 23]
```diff
@@ -0,0 +1,453 @@
...
+	return nil, getPolicyErrorType
+}
+
+func getPolicyBadeDecode(req *http.Request) (*http.Response, error) {
+	header := http.Header{}
+	header.Add("Content-Type", "application/json")
+
+	body := strings.NewReader(`{
+    "name": "readinstance",
+    "id": "95ca7b25-5c8f-4c1b-92da-4276f23807f3",
```
Comment (sean-) [inline identity/policies_test.go]: UUID?

File: identity/policies_test.go [hunk 24]
```diff
@@ -0,0 +1,453 @@
...
+    "rules": [
+      "can listmachine and getmachine"
+    ],
+  }`)
+	return &http.Response{
```
Comment (sean-) [inline identity/policies_test.go]: `\n`

File: identity/policies_test.go [hunk 25]
```diff
@@ -0,0 +1,453 @@
...
+	return nil, deletePolicyErrorType
+}
+
+func updatePolicySuccess(req *http.Request) (*http.Response, error) {
+	header := http.Header{}
+	header.Add("Content-Type", "application/json")
+
+	body := strings.NewReader(`{
+  "name": "Updated Name",
+  "id": "95ca7b25-5c8f-4c1b-92da-4276f23807f3",
```
Comment (sean-) [inline identity/policies_test.go]: UUID?

File: identity/policies_test.go [hunk 26]
```diff
@@ -0,0 +1,453 @@
...
+	return nil, updatePolicyErrorType
+}
+
+func createPolicySuccess(req *http.Request) (*http.Response, error) {
+	header := http.Header{}
+	header.Add("Content-Type", "application/json")
+
+	body := strings.NewReader(`{
+  "name": "Test Policy",
+  "id": "8700e959-4cb3-4337-8afa-fb0a53b5366e",
```
Comment (sean-) [inline identity/policies_test.go]: Different UUID?

File: identity/policies.go (hunk 3 above)
Comment (stack72) [reply identity/policies.go]: We use pointers in all other places - using a pointer here keeps us consistent

File: identity/policies.go (hunk 4 above)
Comment (stack72) [reply identity/policies.go]: We use pointers in all other places - using a pointer here keeps us consistent

File: identity/policies.go (hunk 5 above)
Comment (stack72) [reply identity/policies.go]: We use pointers in all other places - using a pointer here keeps us consistent

File: identity/policies.go (hunk 6 above)
Comment (stack72) [reply identity/policies.go]: We use pointers in all other places - using a pointer here keeps us consistent

**Comment by stack72** [comment]
@sean- this is now ready for review again, the UUIDs in the JSON just represent 2 fake policies I grabbed from the documentation. I updated all of the tests to make sure the correct UUIDs are in place

I also updated the Update test for an error when posting to a different URL the contents of another policy to show an error occurs

P.

**Pull request merged**